# DAV PROJECT
### Faizan ul Haq , Mubashir Zaidi
### i21-1771 , i21-1764
### DS-M  

In [1]:
# !pip install convokit

In [3]:
import convokit as ck
import pandas as pd
import nltk as nltk
import numpy as np
from tqdm import tqdm

import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
import string


In [3]:
# corpus = ck.Corpus(filename=download("movie-corpus"))

In [4]:
corpus = ck.Corpus("movie-corpus")

The Tree Structure of a Corupus:
- Conversations
  - Speakers
    - Utterence by speaker
    
calculating the total amount of values:

In [5]:
corpus.print_summary_stats()

Number of Speakers: 9035
Number of Utterances: 304713
Number of Conversations: 83097


In [16]:
convoDF = corpus.get_conversations_dataframe()
convoDF.head()

vectors meta.movie_idx             meta.movie_name meta.release_year  \
id                                                                           
L1044      []             m0  10 things i hate about you              1999   
L984       []             m0  10 things i hate about you              1999   
L924       []             m0  10 things i hate about you              1999   
L870       []             m0  10 things i hate about you              1999   
L866       []             m0  10 things i hate about you              1999   

      meta.rating meta.votes             meta.genre  
id                                                   
L1044        6.90      62847  ['comedy', 'romance']  
L984         6.90      62847  ['comedy', 'romance']  
L924         6.90      62847  ['comedy', 'romance']  
L870         6.90      62847  ['comedy', 'romance']  
L866         6.90      62847  ['comedy', 'romance']

In [17]:
convoDF.reset_index(inplace=True)           #bring in the index as a column and create a new index column

In [18]:
convoDF.columns

Index(['id', 'vectors', 'meta.movie_idx', 'meta.movie_name',
       'meta.release_year', 'meta.rating', 'meta.votes', 'meta.genre'],
      dtype='object')

In [19]:
convoDF['vectors']

0        []
1        []
2        []
3        []
4        []
         ..
83092    []
83093    []
83094    []
83095    []
83096    []
Name: vectors, Length: 83097, dtype: object

In [20]:
#dropping vectors column as it has no data
convoDF = convoDF.drop(columns=['vectors'])
convoDF.head()

id meta.movie_idx             meta.movie_name meta.release_year  \
0  L1044             m0  10 things i hate about you              1999   
1   L984             m0  10 things i hate about you              1999   
2   L924             m0  10 things i hate about you              1999   
3   L870             m0  10 things i hate about you              1999   
4   L866             m0  10 things i hate about you              1999   

  meta.rating meta.votes             meta.genre  
0        6.90      62847  ['comedy', 'romance']  
1        6.90      62847  ['comedy', 'romance']  
2        6.90      62847  ['comedy', 'romance']  
3        6.90      62847  ['comedy', 'romance']  
4        6.90      62847  ['comedy', 'romance']

In [21]:
release_Years = convoDF['meta.release_year'].unique()
release_Years = np.sort(release_Years)
print(release_Years)

['1927' '1931' '1932' '1933' '1934' '1936' '1937' '1939' '1940' '1941'
 '1942' '1943' '1944' '1945' '1946' '1949' '1950' '1953' '1954' '1955'
 '1956' '1957' '1958' '1959' '1960' '1961' '1963' '1964' '1965' '1966'
 '1967' '1968' '1968/I' '1969' '1970' '1971' '1972' '1973' '1974' '1975'
 '1976' '1977' '1978' '1979' '1980' '1981' '1982' '1983' '1984' '1985'
 '1986' '1987' '1988' '1989' '1989/I' '1990' '1990/I' '1991' '1992'
 '1992/I' '1993' '1994' '1995' '1995/I' '1996' '1996/I' '1997' '1998'
 '1998/I' '1999' '2000' '2000/I' '2001' '2002' '2002/I' '2003' '2003/I'
 '2004' '2004/I' '2005' '2005/I' '2006' '2007' '2007/I' '2008' '2009'
 '2009/I' '2010']


In [22]:
#remove all the years in release_year that have '/I' in the end
release_Years = release_Years[~np.char.endswith(release_Years.astype(str), '/I')]
print(release_Years)


['1927' '1931' '1932' '1933' '1934' '1936' '1937' '1939' '1940' '1941'
 '1942' '1943' '1944' '1945' '1946' '1949' '1950' '1953' '1954' '1955'
 '1956' '1957' '1958' '1959' '1960' '1961' '1963' '1964' '1965' '1966'
 '1967' '1968' '1969' '1970' '1971' '1972' '1973' '1974' '1975' '1976'
 '1977' '1978' '1979' '1980' '1981' '1982' '1983' '1984' '1985' '1986'
 '1987' '1988' '1989' '1990' '1991' '1992' '1993' '1994' '1995' '1996'
 '1997' '1998' '1999' '2000' '2001' '2002' '2003' '2004' '2005' '2006'
 '2007' '2008' '2009' '2010']


In [23]:
#convert years in release year to int
release_Years = release_Years.astype(int)
release_Years

array([1927, 1931, 1932, 1933, 1934, 1936, 1937, 1939, 1940, 1941, 1942,
       1943, 1944, 1945, 1946, 1949, 1950, 1953, 1954, 1955, 1956, 1957,
       1958, 1959, 1960, 1961, 1963, 1964, 1965, 1966, 1967, 1968, 1969,
       1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980,
       1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991,
       1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002,
       2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010])

In [24]:
counts_by_year = (convoDF['meta.release_year'].value_counts())
counts_by_year

1999      6646
1997      5379
2000      5293
1998      5045
2001      4082
          ... 
1969        71
1968/I      54
2004/I      39
2000/I      35
1965        34
Name: meta.release_year, Length: 88, dtype: int64

In [25]:
print(convoDF.shape)

(83097, 7)


### Plan for 1st Visualization
- Now the plan is that we will take out the chunks in "decades"
- The visualization will be a tree and each node represents a decade.
- Every decade chunk node contains genere nodes.
- Every genere chunk node will contain the name of the movies within that Genere. The user has the option to view a word cloud of that movie.
- the user can also view a 1D bar type visualization that shows the most used word/list of words in a single year of that decade

this above matches 83097 number of rows which is the total height of the conversations dataframe, meaning all conversations are available


In [26]:
convoDF.head()

id meta.movie_idx             meta.movie_name meta.release_year  \
0  L1044             m0  10 things i hate about you              1999   
1   L984             m0  10 things i hate about you              1999   
2   L924             m0  10 things i hate about you              1999   
3   L870             m0  10 things i hate about you              1999   
4   L866             m0  10 things i hate about you              1999   

  meta.rating meta.votes             meta.genre  
0        6.90      62847  ['comedy', 'romance']  
1        6.90      62847  ['comedy', 'romance']  
2        6.90      62847  ['comedy', 'romance']  
3        6.90      62847  ['comedy', 'romance']  
4        6.90      62847  ['comedy', 'romance']

In [27]:
#remove all redundant rows from the new_convoDF dataframe
new_convoDF = convoDF.drop(columns=['id'])
new_convoDF = new_convoDF.drop_duplicates()
new_convoDF.reset_index(inplace=True)
new_convoDF.head()


index meta.movie_idx             meta.movie_name meta.release_year  \
0      0             m0  10 things i hate about you              1999   
1    201             m1  1492: conquest of paradise              1992   
2    294             m2                  15 minutes              2001   
3    472             m3       2001: a space odyssey              1968   
4    545             m4                     48 hrs.              1982   

  meta.rating meta.votes                                         meta.genre  
0        6.90      62847                              ['comedy', 'romance']  
1        6.20      10421     ['adventure', 'biography', 'drama', 'history']  
2        6.10      25854           ['action', 'crime', 'drama', 'thriller']  
3        8.40     163227                 ['adventure', 'mystery', 'sci-fi']  
4        6.90      22289  ['action', 'comedy', 'crime', 'drama', 'thrill...

In [28]:
#remove the index column
new_convoDF = new_convoDF.drop(columns=['index'])
new_convoDF.head()

meta.movie_idx             meta.movie_name meta.release_year meta.rating  \
0             m0  10 things i hate about you              1999        6.90   
1             m1  1492: conquest of paradise              1992        6.20   
2             m2                  15 minutes              2001        6.10   
3             m3       2001: a space odyssey              1968        8.40   
4             m4                     48 hrs.              1982        6.90   

  meta.votes                                         meta.genre  
0      62847                              ['comedy', 'romance']  
1      10421     ['adventure', 'biography', 'drama', 'history']  
2      25854           ['action', 'crime', 'drama', 'thriller']  
3     163227                 ['adventure', 'mystery', 'sci-fi']  
4      22289  ['action', 'comedy', 'crime', 'drama', 'thrill...

In [29]:
#convertin the strings in the genere column to lists
import ast
new_convoDF['meta.genre'] = new_convoDF['meta.genre'].apply(lambda x: ast.literal_eval(x))

In [30]:
new_convoDF = new_convoDF[~new_convoDF['meta.release_year'].str.contains(r'\d{4}/I', na=False)]     #remove all the years in release_year that have '/I' in the end
new_convoDF['meta.release_year'] = new_convoDF['meta.release_year'].astype(int)

## making a wordcloud

In [31]:
uttDF = corpus.get_utterances_dataframe()

In [32]:
uttDF.head()

timestamp          text speaker reply_to conversation_id meta.movie_id  \
id                                                                             
L1045      None  They do not!      u0    L1044           L1044            m0   
L1044      None   They do to!      u2     None           L1044            m0   
L985       None    I hope so.      u0     L984            L984            m0   
L984       None     She okay?      u2     None            L984            m0   
L925       None     Let's go.      u0     L924            L924            m0   

                                             meta.parsed vectors  
id                                                                
L1045  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...      []  
L1044  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...      []  
L985   [{'rt': 1, 'toks': [{'tok': 'I', 'tag': 'PRP',...      []  
L984   [{'rt': 1, 'toks': [{'tok': 'She', 'tag': 'PRP...      []  
L925   [{'rt': 0, 'toks': [{'tok': 'Let', 'tag': 'VB'...      []

In [33]:
uttDF.reset_index(inplace=True)           #bring in the index as a column and create a new index column

In [34]:
uttDF.columns

Index(['id', 'timestamp', 'text', 'speaker', 'reply_to', 'conversation_id',
       'meta.movie_id', 'meta.parsed', 'vectors'],
      dtype='object')

In [35]:
uttDF.shape

(304713, 9)

In [36]:
uttDF.head()

id timestamp          text speaker reply_to conversation_id  \
0  L1045      None  They do not!      u0    L1044           L1044   
1  L1044      None   They do to!      u2     None           L1044   
2   L985      None    I hope so.      u0     L984            L984   
3   L984      None     She okay?      u2     None            L984   
4   L925      None     Let's go.      u0     L924            L924   

  meta.movie_id                                        meta.parsed vectors  
0            m0  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...      []  
1            m0  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...      []  
2            m0  [{'rt': 1, 'toks': [{'tok': 'I', 'tag': 'PRP',...      []  
3            m0  [{'rt': 1, 'toks': [{'tok': 'She', 'tag': 'PRP...      []  
4            m0  [{'rt': 0, 'toks': [{'tok': 'Let', 'tag': 'VB'...      []

In [37]:
uttDF.isna().sum()

id                      0
timestamp          304713
text                    0
speaker                 0
reply_to            83097
conversation_id         0
meta.movie_id           0
meta.parsed             0
vectors                 0
dtype: int64

In [38]:
#dropping vectors and timestamp column as it has no data
uttDF = uttDF.drop(columns=['vectors', 'timestamp'])

In [39]:
uttDF.head()

id          text speaker reply_to conversation_id meta.movie_id  \
0  L1045  They do not!      u0    L1044           L1044            m0   
1  L1044   They do to!      u2     None           L1044            m0   
2   L985    I hope so.      u0     L984            L984            m0   
3   L984     She okay?      u2     None            L984            m0   
4   L925     Let's go.      u0     L924            L924            m0   

                                         meta.parsed  
0  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...  
1  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...  
2  [{'rt': 1, 'toks': [{'tok': 'I', 'tag': 'PRP',...  
3  [{'rt': 1, 'toks': [{'tok': 'She', 'tag': 'PRP...  
4  [{'rt': 0, 'toks': [{'tok': 'Let', 'tag': 'VB'...

In [40]:
#count the number of utterences in movie_id m0
uttDF[uttDF['meta.movie_id'] == 'm0'].shape

(669, 7)

In [41]:
new_convoDF.head()

meta.movie_idx             meta.movie_name  meta.release_year meta.rating  \
0             m0  10 things i hate about you               1999        6.90   
1             m1  1492: conquest of paradise               1992        6.20   
2             m2                  15 minutes               2001        6.10   
3             m3       2001: a space odyssey               1968        8.40   
4             m4                     48 hrs.               1982        6.90   

  meta.votes                                meta.genre  
0      62847                         [comedy, romance]  
1      10421    [adventure, biography, drama, history]  
2      25854          [action, crime, drama, thriller]  
3     163227              [adventure, mystery, sci-fi]  
4      22289  [action, comedy, crime, drama, thriller]

In [42]:
#removing all the rows from uttDF that are not in new_convoDF based on the meta.movie_idx column
uttDF = uttDF[uttDF['meta.movie_id'].isin(new_convoDF['meta.movie_idx'])]
uttDF.head()

id          text speaker reply_to conversation_id meta.movie_id  \
0  L1045  They do not!      u0    L1044           L1044            m0   
1  L1044   They do to!      u2     None           L1044            m0   
2   L985    I hope so.      u0     L984            L984            m0   
3   L984     She okay?      u2     None            L984            m0   
4   L925     Let's go.      u0     L924            L924            m0   

                                         meta.parsed  
0  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...  
1  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...  
2  [{'rt': 1, 'toks': [{'tok': 'I', 'tag': 'PRP',...  
3  [{'rt': 1, 'toks': [{'tok': 'She', 'tag': 'PRP...  
4  [{'rt': 0, 'toks': [{'tok': 'Let', 'tag': 'VB'...

In [43]:
uttDF.size        #check if the rows have been removed

2085916

In [44]:
uttDF['meta.parsed'][0]

[{'rt': 1,
  'toks': [{'tok': 'They', 'tag': 'PRP', 'dep': 'nsubj', 'up': 1, 'dn': []},
   {'tok': 'do', 'tag': 'VBP', 'dep': 'ROOT', 'dn': [0, 2, 3]},
   {'tok': 'not', 'tag': 'RB', 'dep': 'neg', 'up': 1, 'dn': []},
   {'tok': '!', 'tag': '.', 'dep': 'punct', 'up': 1, 'dn': []}]}]

### UTTERENCE CLEANING

In [59]:
# # Download stopwords and wordnet if not already downloaded
# nltk.download('stopwords')
# nltk.download('wordnet')

In [47]:
def cleanUtterences(text):
    text = text.replace("'", "")
    words = nltk.word_tokenize(text)
    cleaned_words = []
    
    stop_words = set(stopwords.words('english'))
    
    for word in words:
        # Remove punctuation character by character
        cleaned_word = ''
        for char in word:
            if char not in string.punctuation:
                cleaned_word += char
        if cleaned_word and cleaned_word.lower() not in stop_words:
            cleaned_words.append(cleaned_word)
    
    return cleaned_words


In [ ]:
# Apply the preprocessing function to the 'meta.parsed' column and create a new column 'updated_utt'
uttDF['updated_utt'] = uttDF['text'].apply(cleanUtterences)

In [49]:
uttDF.head()

id          text speaker reply_to conversation_id meta.movie_id  \
0  L1045  They do not!      u0    L1044           L1044            m0   
1  L1044   They do to!      u2     None           L1044            m0   
2   L985    I hope so.      u0     L984            L984            m0   
3   L984     She okay?      u2     None            L984            m0   
4   L925     Let's go.      u0     L924            L924            m0   

                                         meta.parsed  
0  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...  
1  [{'rt': 1, 'toks': [{'tok': 'They', 'tag': 'PR...  
2  [{'rt': 1, 'toks': [{'tok': 'I', 'tag': 'PRP',...  
3  [{'rt': 1, 'toks': [{'tok': 'She', 'tag': 'PRP...  
4  [{'rt': 0, 'toks': [{'tok': 'Let', 'tag': 'VB'...

In [ ]:
uttDF['updated_utt'][0]

''

In [81]:
def cleanMovieNames(text):
    text = text.replace("'", "")
    text = text.replace("`", "")
    text = text.replace("``", "")        
    words = nltk.word_tokenize(text)
    cleaned_words = []
    stop_words = set(stopwords.words('english'))
    
    for word in words:
        cleaned_word = ''
        for char in word:
            if char not in string.punctuation:
                cleaned_word += char
        
        if cleaned_word and cleaned_word not in stop_words:
            cleaned_words.append(cleaned_word)
    
    cleaned_text = ' '.join(cleaned_words)
    
    return cleaned_text

In [64]:
#drop updated_movie_name column if it exists
if 'updated_movie_name' in new_convoDF.columns:
    new_convoDF = new_convoDF.drop(columns=['updated_movie_name'])

In [65]:
#clean the meta.movie_name column
new_convoDF['updated_movie_name'] = new_convoDF['meta.movie_name'].apply(cleanMovieNames)

In [66]:
treeVisualization = {}

for year in tqdm((release_Years)):
    decade = year // 10 * 10
    if decade not in treeVisualization:
        treeVisualization[decade] = {}
    for index, row in new_convoDF.iterrows():
        if row['meta.release_year'] == year:
            for genre in row['meta.genre']:
                if genre not in treeVisualization[decade]:
                    treeVisualization[decade][genre] = []
                utterances = uttDF.loc[(uttDF['meta.movie_id'] == row['meta.movie_idx']) & (uttDF['updated_utt'] != ''), 'updated_utt'].tolist()        #get all the utterances for the movie_id and insert it to a list 
                treeVisualization[decade][genre].append({'movie_name': row['updated_movie_name'], 'utterances': utterances})

  0%|          | 0/74 [00:00<?, ?it/s]

100%|██████████| 74/74 [02:40<00:00,  2.17s/it]


In [67]:
len(treeVisualization[1920]['adventure'][0]['utterances'])

242

In [68]:
#save the nested dictionary as a json file
import json
treeVisualization_fixed = {int(k): v for k, v in treeVisualization.items()}  # convert keys to str
with open('newdata.json', 'w') as fp:
    json.dump(treeVisualization_fixed, fp, indent=4)

## Preprocessing for force directed

In [6]:
pairData = corpus.speaking_pairs(speaker_ids_only=True)

In [7]:
convoDF = corpus.get_conversations_dataframe()

In [8]:
import ast
#reset index
convoDF.reset_index(inplace=True)
#convert meta.genre column to list
convoDF['meta.genre'] = convoDF['meta.genre'].apply(lambda x: ast.literal_eval(x))
convoDF.head()

id vectors meta.movie_idx             meta.movie_name meta.release_year  \
0  L1044      []             m0  10 things i hate about you              1999   
1   L984      []             m0  10 things i hate about you              1999   
2   L924      []             m0  10 things i hate about you              1999   
3   L870      []             m0  10 things i hate about you              1999   
4   L866      []             m0  10 things i hate about you              1999   

  meta.rating meta.votes         meta.genre  
0        6.90      62847  [comedy, romance]  
1        6.90      62847  [comedy, romance]  
2        6.90      62847  [comedy, romance]  
3        6.90      62847  [comedy, romance]  
4        6.90      62847  [comedy, romance]

In [9]:
# Check if the character is already in the nodes list
def is_character_in_nodes(character_name, nodes_list):
    for node in nodes_list:
        if node["character"] == character_name:
            return True
    return False



output={}
output['nodes']=[]
output['links']=[]

for pair in tqdm(pairData):
    id1 , id2 = pair
    speaker1 = corpus.get_speaker(id1)
    speaker2 = corpus.get_speaker(id2)
    movie_id = speaker1.meta['movie_idx']
    gender = speaker1.meta['gender']
    generes = convoDF[convoDF['meta.movie_idx'] == movie_id]['meta.genre'].iloc[0]      #get the movie genre from the movie_idx column
    movie_name = convoDF[convoDF['meta.movie_idx'] == movie_id]['meta.movie_name'].iloc[0]      #get the movie name from the movie_idx column
    movieYear = convoDF[convoDF['meta.movie_idx'] == movie_id]['meta.release_year'].iloc[0]      #get the movie year from the movie_idx column
   #check if movieYear can be converted to int then change movieYear to int and then into a decade
    if movieYear.isnumeric():
        movieYear = int(movieYear)
    else:
        movieYear = movieYear[0:4]
        movieYear = int(movieYear)
   #change the movieyear to decade
    movieYear = movieYear // 10 * 10
    
    # Check for existence before appending
    if not is_character_in_nodes(speaker1.meta['character_name'], output['nodes']):
        temp = {
            "id": speaker1.meta['character_name'],
            "gender": gender,
            "movie": movie_name,
            "year": movieYear,
            "genres": generes
        }
        output['nodes'].append(temp)

    #set freq to be random between 1 and 26
    freq = np.random.randint(1, 26)
    
    temp2={
        "source":speaker1.meta['character_name'],
        "target":speaker2.meta['character_name'],
        "value":freq,
        "movie": movie_name
    }
    output['links'].append(temp2)
    



100%|██████████| 20909/20909 [08:30<00:00, 40.95it/s]


In [12]:
#check if the nodes list has any duplicates
def has_duplicates(listOfNodes):
    for i in tqdm(listOfNodes):
        if listOfNodes.count(i) > 1:
            return True
    return False

has_duplicates(output['nodes'])



100%|██████████| 5347/5347 [00:07<00:00, 743.13it/s]


False

In [26]:
output

{'nodes': [{'character': 'RITA',
   'gender': 'f',
   'movie': 'mulholland dr.',
   'year': 2000,
   'genres': ['drama', 'mystery', 'thriller']},
  {'character': 'LEONARD',
   'gender': 'm',
   'movie': 'a bucket of blood',
   'year': 1950,
   'genres': ['comedy', 'horror']},
  {'character': 'WADE',
   'gender': 'm',
   'movie': 'affliction',
   'year': 1990,
   'genres': ['drama', 'mystery', 'thriller']},
  {'character': 'ROD',
   'gender': '?',
   'movie': "jacob's ladder",
   'year': 1990,
   'genres': ['drama', 'mystery', 'thriller']},
  {'character': 'RUTH',
   'gender': '?',
   'movie': 'titanic',
   'year': 1990,
   'genres': ['drama', 'history', 'romance']},
  {'character': 'JULIA',
   'gender': 'f',
   'movie': 'hellraiser',
   'year': 1980,
   'genres': ['horror']},
  {'character': 'ENZO',
   'gender': 'm',
   'movie': 'le grand bleu',
   'year': 1980,
   'genres': ['adventure', 'drama', 'romance']},
  {'character': 'KANE',
   'gender': 'f',
   'movie': 'citizen kane',
   'ye

In [11]:
#writing the output to a json file
import json
with open('networkDataset.json', 'w') as fp:
    json.dump(output, fp, indent=4)